contracts/secret-storage/SecretManager.sol

```diff
@@ -0,0 +1,601 @@
+        uint256 _secretId,
+        uint256 _index
+    ) internal {
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        if(_index != len - 1)
+            userStorage[_secretId].selectedEnclaves[_index] = userStorage[_secretId].selectedEnclaves[len - 1];
+        userStorage[_secretId].selectedEnclaves.pop();
+    }
+
+    function _refundExcessDepositAndRemoveStore(
+        uint256 _secretId
+    ) internal {
+        address owner = userStorage[_secretId].owner;
+        uint256 remainingDeposit = userStorage[_secretId].usdcDeposit;
+        delete userStorage[_secretId];
+        USDC_TOKEN.safeTransfer(owner, remainingDeposit);
+    }
+
+    function _updateSecretEndTimestamp(
+        uint256 _secretId,
+        uint256 _endTimestamp,
+        uint256 _usdcDeposit,
+        address _owner
+    ) internal {
+        if(userStorage[_secretId].owner != _owner)
+            revert SecretManagerNotUserStoreOwner();
+
+        if(_endTimestamp < block.timestamp)
+            revert SecretManagerInvalidEndTimestamp();
+
+        uint256 currentEndTimestamp = userStorage[_secretId].endTimestamp;
+        if(block.timestamp > currentEndTimestamp)
+            revert SecretManagerAlreadyTerminated();
+
+        _checkIfSecretAcknowledged(_secretId);
+
+        if(_endTimestamp > currentEndTimestamp) {
+            USDC_TOKEN.safeTransferFrom(_owner, address(this), _usdcDeposit);
+            userStorage[_secretId].usdcDeposit += _usdcDeposit;
+
+            uint256 minUsdcDeposit = (_endTimestamp - userStorage[_secretId].startTimestamp) * userStorage[_secretId].sizeLimit * SECRET_STORE_FEE_RATE * NO_OF_NODES_TO_SELECT;
+            _checkUsdcDeposit(userStorage[_secretId].usdcDeposit, minUsdcDeposit);
+        }
+        else {
+            uint256 removedDuration = currentEndTimestamp - _endTimestamp;
+            uint256 usdcRefund = removedDuration * userStorage[_secretId].sizeLimit * SECRET_STORE_FEE_RATE * NO_OF_NODES_TO_SELECT;
+            
+            userStorage[_secretId].usdcDeposit -= usdcRefund;
+            USDC_TOKEN.safeTransfer(_owner, usdcRefund);
+        }
+
+        userStorage[_secretId].endTimestamp = _endTimestamp;
+
+        emit SecretEndTimestampUpdated(_secretId, _endTimestamp);
+    }
+
+    function _checkIfSecretAcknowledged(
+        uint256 _secretId
+    ) internal view {
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        for (uint256 index = 0; index < len; index++) {
+            if(!userStorage[_secretId].selectedEnclaves[index].hasAcknowledgedStore)
+                revert SecretManagerUnacknowledged();
+        }
+    }
+
+    function _terminateSecret(
+        uint256 _secretId
+    ) internal {
+        uint256 noOfSelectedEnclaves = userStorage[_secretId].selectedEnclaves.length;
+        _updateUsdcDeposit(_secretId, noOfSelectedEnclaves);
+        _refundExcessDepositAndRemoveStore(_secretId);
+
+        emit SecretTerminated(_secretId);
+    }
```

Comment:
add check for the owner

Currently this method can be used for two purposes
1. Terminating an active secret store
2. Getting refund for already expired secret store

we should have different functions for them, with different restriction as following:
(1) can only be done by owner before the expiry
(2) can be done by anyone after the expiry

Agree?

Reply:
yeah seems good